bindings_ffi/src/crypto.rs

```diff
@@ -0,0 +1,167 @@
+use thiserror::Error;
+use xmtp_cryptography::ethereum;
+
+#[derive(Debug, Error, uniffi::Error)]
+pub enum FfiCryptoError {
+    #[error("invalid length")]
+    InvalidLength,
+    #[error("invalid key")]
+    InvalidKey,
+    #[error("signing failure")]
+    SignFailure,
+    #[error("pubkey decompress failure")]
+    DecompressFailure,
+}
+
+impl From<ethereum::EthereumCryptoError> for FfiCryptoError {
+    fn from(err: ethereum::EthereumCryptoError) -> Self {
+        match err {
+            ethereum::EthereumCryptoError::InvalidLength => FfiCryptoError::InvalidLength,
+            ethereum::EthereumCryptoError::InvalidKey => FfiCryptoError::InvalidKey,
+            ethereum::EthereumCryptoError::SignFailure => FfiCryptoError::SignFailure,
+            ethereum::EthereumCryptoError::DecompressFailure => FfiCryptoError::DecompressFailure,
+        }
+    }
+}
+
+/// 1) Ethereum compatible public key from 32-byte private key.
+///    Returns **65-byte uncompressed** (0x04 || X || Y)
+#[uniffi::export]
+fn ethereum_generate_public_key(private_key32: Vec<u8>) -> Result<Vec<u8>, FfiCryptoError> {
+    let public_key = ethereum::public_key_uncompressed(&private_key32)?;
+    Ok(public_key.to_vec())
+}
+
+/// 2) Ethereum recoverable signature.
+///    Returns **65 bytes r||s||v**, with **v in {0,1}** (parity bit).
```

Comment:
We say v is `{0, 1}` here, but we test below and assert that it is `{27,28}`.

This has tripped us up before. We should loudly call out that it will be 27 or 28 if that's the expected return. Also need to make sure it matches what the old library was doing

Reply:
@neekolas Just checked this. Turns out xmtp-ios was the exception here, in that it currently uses recovery_id 0: https://github.com/xmtp/xmtp-ios/blob/f88ab0280203f53eee5b32de5d7aeb0b5a984c1d/Sources/XMTPiOS/KeyUtil.swift#L104

- Android uses web3cryptoj which adds 27 to the recId
- xmtp-js uses viem which adds the 27 to the recId
- alloy in libxmtp also adds the 27 to recId

So I'm inclined to update ios to be compatible with the rest of our test code here and just make the comments call that out loudly in xmtp-ios. Updated the comment here for clarity as well.

Reply:
Love consistency. Makes sense to have everything be the same.

Pretty sure this whole dumb mess of non-standard recovery IDs is Satoshi's fault.